test/Microsoft.ML.Tests/Scenarios/Scenario3_SentimentPrediction.cs

```diff
@@ -23,7 +24,32 @@ public void TrainAndPredictSentimentModelTest()
         {
             string dataPath = GetDataPath(SentimentDataPath);
             var pipeline = new LearningPipeline();
-            pipeline.Add(new TextLoader<SentimentData>(dataPath, useHeader: true, separator: "tab"));
+
+            pipeline.Add(new TextLoader(dataPath)
+            {
+                Arguments = new TextLoaderArguments
```

Comment:
I thought we were going to go with the API that was spec'd in [https://github.com/dotnet/machinelearning/blob/master/Documentation/specs/mvp.md#training](https://github.com/dotnet/machinelearning/blob/master/Documentation/specs/mvp.md#training)?

```
var options = CsvOptions.CreateFrom<HousePriceData>(separator: '\t', header: true);


namespace Microsoft.MachineLearning
{
    public enum DataType
    {
        Boolean,
        Int32,
        Single,
        Double,
        String        
    }
    public class CsvColumn
    {
        public CsvColumn(string name, DataType dataType, int ordinal);
        public string Name { get; }
        public DataType DataType { get; }
        public int Ordinal { get; }
    }
    public class CsvOptions
    {
        public static CsvOptions CreateFrom(char separator = ',', bool hasHeader = true, Type rowType);
        public static CsvOptions CreateFrom<T>(char separator = ',', bool hasHeader = true);

        public CsvOptions();
        public char Separator { get; set; }
        public bool HasHeader { get; set; }
        public CsvColumnCollection Columns { get; }
    }
    public class CsvColumnCollection : Collection<CsvColumn>
    {
        public CsvColumn Add(string name, DataType dataType);
        public CsvColumn Add(string name, DataType dataType, int ordinal);
    }
}
```

Do we still have plans on implementing this API?

Reply:
On this format, should the types be just `Type`? How about locales while reading or writing? Should these separate CSV specific types such as `CsvColumn` and `CsvColumnCollecttion` for CSV data be rolled to some more generic structure? I mean, it would feel plausible to think it's just a collection of data in row or column format (or some other format) whereas any data read in any format (JSON, XML, whatnot). This is a simple example (but not that good since not heterogenous): https://gist.github.com/veikkoeeva/50c8f38ec46b0a3ce70467d16af00dc1, then ADO.NET is fuller taking into account types too: https://docs.microsoft.com/en-us/dotnet/api/system.data.datatable?view=netframework-4.7.2 (https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/ado-net-datasets). I don't advocate taking a dependecy on ADO.NET, but might have food for thoughts.

In general readong or writing CSV is "easy" as https://stackoverflow.com/questions/5116604/read-csv-using-linq, in practice there are locale format spefic things (and just problems) and one might consider taking a dependency to a library (i.e. https://www.nuget.org/packages?q=CSV) and then having, say, an extension method such as `.FromCsv(this FileInfo` (and other appropriate convenience overloads).

Reply:
@veikkoeeva the types being a type instead of an `enum` might be a good idea -- maybe something along the lines of a helper create method. I'd still like to have an `enum` in any such convenience, since it becomes clear what types are supported.

Regarding your other notes about more types, I certainly agree with that, but we have elsewhere not addressed these problems by changing `TextLoader`, but rather by introducing new implementations of `IDataLoader` (in the case where we're reading new formats, see e.g., [PR 61](https://github.com/dotnet/machinelearning/pull/61)), or just an `IDataView` in the case that nothing really is being loaded per se (see e.g., [PR 106](https://github.com/dotnet/machinelearning/pull/106), where ). Something like ADO.NET you could imagine having it, phrasing it as an `IDataView` implementation, then just plugging it in, and it should work. Of course, someone would have to write that implementation. (The name `TextLoader` is somewhat unfortunate, since it has, I've noticed, created the impression that it is meant to read any conceivable type of text, rather than conforming to its own specific format.)

@eerhardt thanks for sharing this. If the suggestion was we take this document as a guideline, then we should consider this PR or more precisely its successor [PR 142](https://github.com/dotnet/machinelearning/pull/142) as perhaps an attempt to generate something in that spirit in that it solves the problem, but that being an actual attempt to solve the problem and undergoing peer review will necessarily diverge from a document like this.

I get the sense though that you're thinking this proposal should be taken literally. I also get the sense this proposal was written somewhat hurriedly. The following three flaws occur to me on cursory examination: Ordinal as `int` whiffed on vector valued columns, `DataType` needlessly duplicates the existing concept of `DataKind` (but less completely, it missed many types including those related to dates and times, most seriously, which people have already opened issues pointing out), and finally has `Csv` in its name despite the fact the separator/delimiter is configurable. Anyway, perhaps open an issue, these and other problems can . (Mayhap one is already opened on this subject, I don't know.)

Reply:
You could also have a public collection `SupportedTypes` and have a check in construction that throws an error with appropriate error message if unsupported types are tried. The benefit of `Type` is in all the operations it's clear what is the type (and its semantics) after it has been constructed. I.e., checks and transformatios only in source -- possibly in the sink -- and plays nicely with the rest of the ecosystem.

By the way, if you look at ADO.NET, you notice there are vendor specific types and then just general ADO.NET types and there's some sort of a mapping between them when necessary. It looks like [this](https://github.com/dotnet/orleans/blob/51853385739714409db9fb8f787a54ec8c9fd5e2/src/AdoNet/Shared/Storage/DbExtensions.cs#L33). Notably there it's known that's all that's needed, but a more general purpose mapping would take the vendor invariant and use that to map types better (Oracle, for instance, doesn't have `bool` type).

In a way it looks like this is going into the direction of ADO.NET. If you look at [here](https://blogs.msdn.microsoft.com/adonet/2012/07/15/using-sqldatareaders-new-async-methods-in-net-4-5-part-2-examples/), it's like opening a source, reading it, transforming the results into some other format, and then processing it in some way. (And you can wrap that one like [here](https://github.com/dotnet/orleans/blob/51853385739714409db9fb8f787a54ec8c9fd5e2/src/AdoNet/Shared/Storage/RelationalStorage.cs#L214) and then offer a simple API like [here](https://github.com/dotnet/orleans/blob/51853385739714409db9fb8f787a54ec8c9fd5e2/src/AdoNet/Shared/Storage/RelationalStorageExtensions.cs#L186) (see `ReflectionSelector` that uses the mapping referred earlier) and by playing that idea, maybe many source could be handle with that by supplying an lambda).

I apologize if it's exhaustive I push various view points. I suspect, though, you need to explain these things in any event later to people asking -- at least if MachineLearning.NET becomes popular as we hope. :) I'd like to ask also (and did ponder in Gitter shortly), if your're building a composition pipeline with reflection, and want to use someting less fragile (judging from the comments), would [MEF](https://docs.microsoft.com/en-us/dotnet/framework/mef/) be an option? And if it would, then maybe using dependency injection like much of other frameworks and .NET ecosystem? And if not, maybe there needs to be documentation explaining how to do similarish things.